### ISSUE TITLE
Non-positive-definite forecast error covariance matrix encountered at period

### ISSUE THREAD

**Issue by samanvayagarwal** [statsmodels/statsmodels#4333]
Hi,

I am a beginner in tsa package with python. I am trying to estimate the following model using statespace methods - 
y(t) = alpha*x(t) + v(t)
c(t) = x(t) + u(t)
x(t) = x(t-1) + e(t)

I first simulated a sample and defined the class to call the MLEModel module but I get an error that says "Non-positive-definite forecast error covariance matrix encountered at period 0" and I can't figure out why. Please see below the complete code. Any help will be greatly appreciated!

Thanks!

`import statsmodels.tsa.statespace as sp
import numpy as np
from statsmodels.tsa.statespace.tools import (constrain_stationary_univariate, unconstrain_stationary_univariate)

# simulate sample

alpha, sigmasqv, sigmasqu, sigmasqe = 1, 0.5, 1.5, 1
T = 500
burn =  T/2

einit = sigmasqe**0.5*np.random.randn(T+burn)
uinit = sigmasqu**0.5*np.random.randn(T+burn)
vinit = sigmasqv**0.5*np.random.randn(T+burn)
xinit, cinit, yinit = np.zeros(T+burn), np.zeros(T+burn), np.zeros(T+burn)
xinit[0] = einit[0]

for t in range(1,T+burn):
    xinit[t] = xinit[t-1] + einit[t]

cinit = xinit + uinit
yinit = alpha*xinit + vinit
y, c, x = yinit[burn:], cinit[burn:], xinit[burn:]

simdata = np.concatenate((y,c), axis=0).reshape(2,T)

# Construct the model
class simplemodel(sp.mlemodel.MLEModel):
    start_params = [0, 1, 1, 1.5]
    param_names = ['alpha', 'sigma.v', 'sigma.u', 'sigma.e']

    def __init__(self, endog):
        # Initialize the state space model
        super(simplemodel, self).__init__(endog, k_states=1, k_posdef=1)

        # Setup the fixed components of the state space representation
        #self['design', 0, 0] = [[0.0], [1.0]]
        self['design', 1, 0] = 1.
        self['transition', 0, 0] = 0.
        self['selection', 0, 0] = 1.
        self.initialize_approximate_diffuse()
        self.loglikelihood_burn = 1

    def transform_params(self, params):
        alpha = constrain_stationary_univariate(params[0:1])
        sigmasq = constrain_stationary_univariate(params[1:4]**2)
        return np.r_[alpha, sigmasq]

    def untransform_params(self, params):
        alpha = unconstrain_stationary_univariate(params[0:1])
        sigma = unconstrain_stationary_univariate(params[1:4]**0.5)
        return np.r_[alpha, sigma]

    # Describe how parameters enter the model
    def update(self, params, **kwargs):
        params = super(simplemodel, self).update(params, **kwargs)

        self['design', 0, 0] = params[0]
        self['obs_cov', 0, 0] = params[1] 
        self['obs_cov', 1, 1] = params[2]
        self['state_cov', 0, 0] = params[3]
        

model = simplemodel(simdata)
res = model.fit()
print res.summary()`

**Comment by ChadFulton** [comment]
You're very close, there are just two things that are going wrong:

1. The base state space classes assume that the data is being passed in the shape `(nobs, k_endog)` and your `simdata` variable is shaped `(2, 500)`, so the model is currently assuming that you have 500 variables observed at two points. Instead, just pass in `simdata.T` or similar.

2. In your `transform_params` and `untransform_params` methods, you're applying `constrain_stationary_univariate` and `unconstrain_stationary_univariate` to the loading alpha and all of the variances, which you shouldn't be doing (at least as far as I can tell there's no reason you'd want to do that). I assume that you're intending to not restrict alpha at all, but you want to force the variances to be positive. In that case, the methods should be:

```python
    def transform_params(self, params):
        alpha = params[0:1]
        sigmasq = params[1:4]**2
        return np.r_[alpha, sigmasq]

    def untransform_params(self, params):
        alpha = params[0:1]
        sigma = params[1:4]**0.5
        return np.r_[alpha, sigma]
```

I made those two changes and the model worked for me. Let me know how it goes for you.

**Comment by samanvayagarwal** [comment]
Thanks Chad! It worked. I realized my mistake in transform functions but was confused about the shape of input data. I guess I am still trying to understand how the algorithm works but looks like it returns variances and not standard deviations so my names should be consistent with that too. 

Btw, Thank you for you the resources on your website. They are really helpful 👍
